docs/Threading.md

```diff
@@ -0,0 +1,171 @@
+Each of these threads calls `lthread_run` to start an instance of the lthread scheduler.
+
+The ethreads do not use the FS segment on x86-64.
+The GS segment stores a pointer to the Open Enclave state and the lthread scheduler.
+The [lthread context switch](https://github.com/lsds/sgx-lkl/blob/796b346d4762a93be7bfc2ba5ce97b9ab840a4bd/src/sched/lthread.c#L466) [switches the FS segment register](https://github.com/lsds/sgx-lkl/blob/796b346d4762a93be7bfc2ba5ce97b9ab840a4bd/src/sched/lthread.c#L411) and makes this available for higher layers in the threading stack.
+
+LThreads
+--------
+
+The lthread library provides a *cooperative* threading model.
+The code lives in the [src/sched](../src/sched) directory in the SGX-LKL repository.
+
+### The main scheduler loop
+
+The main loop for the scheduler is in the [`lthread_run`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L214) function.
+The function [pulls the next thread from the run queue](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L234) and then [calls `_lthread_resume`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L242).
+
+The [`_lthread_resume`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L453) function does most of the work for running a thread.
+`_lthread_resume` sets up the initial thread state if required and then calls [`_switch`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L482), a short assembly routine that saves all registers that are preserved across function calls and restores the same set of registers from a buffer.
+The `_switch` routine [restores the stack pointer](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L147) of the thread being switched to and pushes the new instruction pointer value onto the new stack.
+
+If the thread is a new thread, its instruction pointer was [set during thread initialisation](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L540).
+If the thread has previously yielded, the instruction pointer is the [return address for the `_switch` call that switched away from the lthread](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L134).
+
+Note that `_switch` does *not* switch between two arbitrary lthreads.
+One of the threads involved in a `_switch` call is always the scheduler.
+The `_switch` call in `_lthread_resume` switches to another thread, the call [in `_lthread_yield_cb`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L340) and [in `_lthread_yield`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L346) switch back to the scheduler.
+
+
```

Comment:
```suggestion
After the execution is switched back to the scheduler, `lthread_run` checks for virtio wake up events and might add a sleeping lthread to the run queue. 

`lthread_run` maintains a contiguous idle loop count. In each loop, if no lthread is dequeued from the run queue and no sleeping thread is enquued to the run queue,
the loop is considered as a idle loop. Any non-idle loop resets the contiguois idele loop count. Once the contiguous idle loop count reaches a set threshold,
`lthread_run` invokes an OE OCALL to put the ethread on sleep for a set period. The ethread will be resumed by the host after the sleep period passes.

`lthread_run` breaks out of the loop only when the enclave is terminating.
```

Reply:
Reformatting (one sentence per line, so future diffs are not so painful) and correcting slightly:

```suggestion
After the running lthread yeidds, `lthread_run` checks whether any sleeping threads (those blocked waiting for event channels or futexes) are runnable and, if so, adds them to the queue.

`lthread_run` maintains a count of consecutive loop iterations in which there were not runnable lthreads.
Once this reaches a threshold, the scheduler issues an ocall that suspends execution of the ethread until either an event channel is signaled or a timeout expires.

`lthread_run` exits the loop and returns only when the enclave is terminating.
```